### DISCUSSION TITLE
How to utilize the lifetime of `&mut Parts` when implementing `FromRequestParts`.

### DISCUSSION THREAD

**Discussion by twitchax** [tokio-rs/axum#2189] [Q&A] [answered]
Essentially, I want to capture the `&str` value of a header without doing any clones.  Something like this.

```rust
/// A request guard to get Last-Modified data.
#[derive(Debug)]
pub struct IfModifiedSince<'r>(&'r str);

#[async_trait]
impl<'r, S> FromRequestParts<S> for IfModifiedSince<'r>
where
    S: Send + Sync,
{
    type Rejection = (StatusCode, &'static str);

    async fn from_request_parts(parts: &mut Parts, state: &S) -> Result<Self, Self::Rejection> {
        let maybe_lm = parts.headers.get("If-Modified-Since").map(|h| h.to_str().unwrap()).map(IfModifiedSince);

        match maybe_lm {
            Some(lm) => Ok(lm),
            None => Ok(IfModifiedSince("NONE")),
        }
    }
}
```

Here is the error, which makes sense.

<img width="1004" alt="image" src="https://github.com/tokio-rs/axum/assets/549523/88591664-1dd8-465a-af13-882077d45417">

Since the lifetimes are on the _function_, there is no way to cleanly indicate that I want the lifetime of my holder struct to match one of the lifetimes of the implementation (i.e., `&mut Parts`).  Is there any way to resolve this short of axum moving the lifetimes to the trait?

FWIW, this is how Rocket does it (lifetime on the trait), and it allows defining a custom extractor that can be tied to the lifetime of the request without cloning.

```rust
#[async_trait]
impl<'r> FromRequest<'r> for IfModifiedSince<'r> {
    type Error = &'static str;

    async fn from_request(request: &'r Request<'_>) -> Outcome<Self, Self::Error> {
        let maybe_lm = request.headers().get_one("If-Modified-Since").map(IfModifiedSince);

        match maybe_lm {
            Some(lm) => Outcome::Success(lm),
            None => Outcome::Success(IfModifiedSince("NONE")),
        }
    }
}
```

**Comment by jplatte** [comment, marked as answer]
You can't borrow from the request in handlers with the current trait, and axum intentionally doesn't use lifetime generics in there, partly because that would just make the trait harder to grasp for users. But also, the current design has the `FromRequest` trait consume the request, which is incompatible with keeping borrows to the request around for the handler function.

The only thing you *can* do to avoid cloning is to take ownership of the header by removing it from the headers (which means it won't be available for subsequent extractors).

**Reply by twitchax** [reply]
Hmmm, interesting.  The `FromRequest` argument makes sense.  It's not a big deal, but I liked that quality of not need to clone when using Rocket.

**Reply by jplatte** [reply]
What you can do instead is *parse* the header in the extractor and only allocate as much as required to get an owned parsed header. This is essentially what [`TypedHeader`](https://docs.rs/axum/latest/axum/struct.TypedHeader.html) does. For some headers like `age`, this even means no allocation.
